V4 - docs: Revamp the QuantumComputer and the WavefunctionSimulator sections. (#1643)
rigetti/pyquil: docs/source/the_quantum_computer.rst

bugs. To learn more about how to add noise models to your virtual ``QuantumComputer`` instance, check out
:ref:`noise`.

Differences between a QVM and a QPU based ``QuantumComputer``
=============================================================

As mentioned above, pyQuil is designed such that code based on a ``QuantumComputer`` can be used in more or less the same way,

> regardless if it is based on a QVM or QPU. However, depending on which you are using, the subcompoments have additional features

Reviewer: ```suggestion
regardless of whether it is based on a QVM or QPU. However, depending on which you are using, the subcompoments have additional features
```

worth knowing about.

For instance, if your code targets a QVM, ``qc.qam`` will be a :py:class:`~pyquil.api.QVM``` instance, and ``qc.compiler`` will
be a :py:class:`~pyquil.api.QVMCompiler` instance. However, if your code targets a QPU, ``qc.qam`` will be a :py:class:`~pyquil.api.QPU` instance, and ``qc.compiler`` will be a :py:class:`~pyquil.api.QPUCompiler` instance.

While these subcomponents follow common interfaces, namely :py:class:`~pyquil.api.QAM` and
:py:class:`~pyquil.api.AbstractCompiler`, there may be some methods or properties that are accessible on the QPU-based instances
but not on the QVM-based instances, and vice versa.

> You can access these features and keep your code robust by performing type checks on ``qc.qam`` and/or ``qc.compiler``.

Reviewer: Good tip.

Author: Lifted from the GitBook, possibly a tip you wrote long ago? 😂

For example, if you wanted to refresh the calibration program, which only applies to QPU-based ``QuantumComputers``, but still
want a script that works for both QVM and QPU targets, you could do the following:

.. testcode:: differences

    from pyquil import get_qc
    from pyquil.api import QPUCompiler

    qc = get_qc("2q-qvm")  # or "Aspen-M-3" 

    if isinstance(qc.compiler, QPUCompiler):
        # Working with a QPU - refresh calibrations
        qc.compiler.get_calibration_program(force_refresh=True)

Providing your own quantum processor topology
=============================================

> You can provide your own quantum processor topology by specifying qubits as a number, and which edges exist
> between those qubits. Here is an example that uses a subset of the instruction set architecture of a

Reviewer: ```suggestion
You can provide your own quantum processor topology by specifying qubits (as numeric indices) and edges.
Here is an example that uses a subset of the instruction set architecture of a
```
